src/librustc_mir/interpret/snapshot.rs

```diff
@@ -46,361 +85,423 @@ pub(crate) struct InfiniteLoopDetector<'mir, 'tcx> {
+            && block == &other.block
+            && stmt == &other.stmt
+            && span == &other.span
+            && locals.iter().map(IgnoreAllocId)
+                .eq(other.locals.iter().map(IgnoreAllocId))
     }
 }
 
-impl<'a, 'tcx, Ctx> Snapshot<'a, Ctx> for &'a LocalState<'tcx>
-    where Ctx: SnapshotContext<'a>,
-{
-    type Item = LocalValue<(), AllocIdSnapshot<'a>>;
+impl<'mir, 'tcx> Hash for IgnoreAllocId<'_, Frame<'mir, 'tcx>> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // See `PartialEq` impl above for description of ignored fields.
+        let super::Frame {
+            body,
+            instance,
+            locals,
+            block,
+            stmt,
+            span,
+            return_to_block: _,
+            return_place: _,
+            extra: (),
+        } = &self.0;
 
-    fn snapshot(&self, ctx: &'a Ctx) -> Self::Item {
-        let LocalState { value, layout: _ } = self;
-        value.snapshot(ctx)
+        std::ptr::hash::<mir::Body<'tcx>, _>(*body, state);
+        instance.hash(state);
+        block.hash(state);
+        stmt.hash(state);
+        span.hash(state);
+
+        for local in locals {
+            IgnoreAllocId(local).hash(state);
+        }
     }
 }
 
-impl<'b, 'mir, 'tcx> SnapshotContext<'b>
-    for Memory<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>>
-{
-    fn resolve(&'b self, id: &AllocId) -> Option<&'b Allocation> {
-        self.get_raw(*id).ok()
+impl PartialEq for IgnoreAllocId<'_, LocalState<'tcx>> {
+    fn eq(&self, IgnoreAllocId(other): &Self) -> bool {
+        // This *must* remain exhaustive to ensure that all fields are considered for equality.
+        let super::LocalState {
+            value,
+            layout: _, // Memoized result of the layout query.
+        } = self.0;
+
+        match (value, &other.value) {
+            // Compare both `Operand`s with any `AllocId`s erased.
+            (LocalValue::Live(op), LocalValue::Live(other_op))
+                => op.erase_alloc_id() == other_op.erase_alloc_id(),
+
+            // Otherwise we can use the `PartialEq` impl of `LocalValue` directly.
+            _ => value == &other.value,
+        }
     }
 }
 
-/// The virtual machine state during const-evaluation at a given point in time.
-/// We assume the `CompileTimeInterpreter` has no interesting extra state that
-/// is worth considering here.
-#[derive(HashStable)]
-struct InterpSnapshot<'mir, 'tcx> {
-    // Not hashing memory: Avoid hashing memory all the time during execution
-    #[stable_hasher(ignore)]
-    memory: Memory<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>>,
-    stack: Vec<Frame<'mir, 'tcx>>,
+impl Hash for IgnoreAllocId<'_, LocalState<'tcx>> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // See `PartialEq` impl above for description of ignored fields.
+        let super::LocalState {
+            value,
+            layout: _,
+        } = self.0;
+
+        match value {
```

Comment:
you also need to hash the discriminant when you hash an enum's fields, otherwise the `Live` variant may end up with the same hash as a `Dead` or `Uninitialized` variant (without being a true has collision).